Migrate RepositoriesMetadata to ProjectCustom

ywangd opened the pull request:
This PR migrates RepositoriesMetadata from Metadata#ClusterCustom to Metadata#ProjectCustom and handles wire BWC.

Resolves: ES-10477

ywangd committed:
wip

ywangd committed:
diff and tests

ywangd committed:
fix test

ywangd committed:
Merge remote-tracking branch 'origin/main' into mp/repositories-metadata-migration

ywangd committed:
fix tests

ywangd committed:
explicit default project for repo service

ywangd committed:
fix tests

ywangd committed:
comments and renames

ywangd committed:
toxcontent tests and fixes

ywangd committed:
bwc tests with version before multi-project

ywangd committed:
rephrase

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

ywangd committed:
Merge remote-tracking branch 'origin/main' into mp/repositories-metadata-migration

ywangd commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -2399,7 +2399,7 @@ private ClusterState getClusterStateWithUpdatedRepositoryGeneration(ClusterState
             .metadata(
                 Metadata.builder(currentState.getMetadata())
-                    .putCustom(
+                    .putDefaultProjectCustom(
```
This makes the repository explicitly work with the default project. Otherwise, it throws MultiProjectPendingException in some multi-project enabled tests. It worked previously because repositories are cluster custom which does not check the number of projects. It is a temporary workaround (the method is marked as deprecated for removal) and will be removed once repository and snapshot fully support multi-project.

ywangd commented on x-pack/plugin/core/src/test/java/org/elasticsearch/license/LicensesMetadataSerializationTests.java:
```diff
@@ -64,14 +64,14 @@ public void testXContentSerializationOneSignedLicenseWithUsedTrial() throws Exce
         LicensesMetadata licensesMetadata = new LicensesMetadata(license, TrialLicenseVersion.CURRENT);
-        RepositoryMetadata repositoryMetadata = new RepositoryMetadata("repo", "fs", Settings.EMPTY);
-        RepositoriesMetadata repositoriesMetadata = new RepositoriesMetadata(Collections.singletonList(repositoryMetadata));
+        NodesShutdownMetadata nodesShutdownMetadata = new NodesShutdownMetadata(Map.of());
```
This test just needs a different Metadata#ClusterCustom that is not licenses.

DaveCTurner reviewed (commented):
I'm pretty concerned with the complexity of the surgery being introduced into the (already complex) diff-serialization process. Did we explore any alternatives? For instance, could we keep this info in cluster-wide customs until `org.elasticsearch.cluster.ClusterState#getMinTransportVersion` advances far enough that we won't need the BwC complexity?

ywangd commented:
It would require v10 to completely advance this change since v9.0 still stores it in cluster customs. I'd rather to pay for the complexity once similar to how we did it for persistent tasks split so that what's left is removing it once we hit v10, part of it can even be dropped once serverless advances in a few week's time. And it will be in the "final" state. I am sure v10 will come with its own pile of tasks, so overall I think it's better to fix it now.

DaveCTurner commented:
I don't think that's a very strong argument for introducing this level of complexity right now. We will never need to send a diff from a v9.0 node (using cluster-level repositories) to any v10.x node using project-level ones, and we will need to keep the code for fixing this up when loading the `Metadata` from disk all the way through the 10.x series either way to support a full-cluster-restart upgrade between these versions.

DaveCTurner commented:
> part of it can even be dropped once serverless advances in a few week's time

Can you expand on this? What exactly will be dropped? Sorry if I've missed something obvious but I couldn't work out which bit you meant here.

ywangd commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1016,15 +1066,103 @@ public void writeTo(StreamOutput out) throws IOException {
+            MapDiff<ProjectId, ProjectMetadata, Map<ProjectId, ProjectMetadata>> multiProject,
+            MapDiff<String, ReservedStateMetadata, ImmutableOpenMap<String, ReservedStateMetadata>> reservedStateMetadata
+        ) throws IOException {
+            assert out.getTransportVersion().onOrAfter(TransportVersions.MULTI_PROJECT)
+                && out.getTransportVersion().before(TransportVersions.REPOSITORIES_METADATA_AS_PROJECT_CUSTOM) : out.getTransportVersion();
```
@DaveCTurner I meant this method which is most of the new complexity. Versions between `MULTI_PROJECT` and `REPOSITORIES_METADATA_AS_PROJECT_CUSTOM` are only released in serverless. Once serverless rolls forward, we can remove it.

ywangd commented:
Diffs are not needed for full cluster restart, only xcontent which has rather minimal BWC changes.

DaveCTurner commented:
I see, ok, I had missed that if this lands in 9.1 then that becomes dead code. I see there is already prior art for this diff-surgery approach too. I didn't know that, and FTR my discomfort with this PR extends to that prior art too, but I guess we have to roll forwards with it for now.

pxsalehi reviewed (commented):
To my eyes, these changes look ok. But this is quite tricky to review. So I'd rather leave the final review/LGTM to David.

ywangd commented:
Thanks for taking a look, Pooya. Let me provide a bit high level explanation and see whether it helps. 

The issue is that we need to split an entry from `Metadata#customs` and move it into `ProjectMetadata#customs`. We did similar split before merging back to main (`TransportVersions.MULTI_PROJECT`, hereafter referred as `MP`) and this is part of the prior art. For this new split, we can reuse the logic if the old version is before `MP`. The complexity is mostly for communicating with the versions on-or-after `MP`. 

BWC reading and writing for the entire `Metadata` are not bad at all. The challenge comes from the Diffs due to `MapDiff` manipulations and how nested it can be. If we look past the `MapDiff` manipulation, conceptually, the logic is rather straightforward. For example, the steps for writing a Diff is as follows:
1. Iterate through the map of `ProjectMetadataDiff` and separate `RepositoriesMetadata` (or its diff) from the project level customs. Note that we throw if anything is found for non-default project.
2. Combine `RepositoriesMetadata` (or its diff) separated in step 1 with the cluster level customs. 
3. Send the combined cluster level customs and the processed map of `ProjectMetadataDiff` from step 1.

The above logic is encapsulated in method `writeDiffWithRepositoriesMetadataAsClusterCustom` which contributes to most of the new complexity (reading from old versions is a lot simpler due to merging into `MapDiff` being easier). Fortunately, as [commented earlier](https://github.com/elastic/elasticsearch/pull/125398#discussion_r2009632097), this method will soon become unused once Serverless deployments roll forward. So the complexity does _not_ stay with us for very long. This BWC logic is tested with both [unit tests](https://github.com/elastic/elasticsearch/pull/125398/files#diff-f900145269f33a81930fe3a9a151493cca6fd25b863a14d178f66828f37e7b4c) and the serverless rolling upgrade test in the linked PR. Overall I think it is safe to proceed.

@tvernum @nielsbauman Though it is labelled with snapshot/restore, the changes are more for namespacing and tied with previous Metadata changes. I'd appreciate if you could also take a look at this PR since you are familiar with the context. Thank you!

nielsbauman commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1016,15 +1066,103 @@ public void writeTo(StreamOutput out) throws IOException {
+                    throwForVersionBeforeRepositoriesMetadataMigration(out);
                 }
+                // RepositoriesMetadata found for the default project as an upsert, package it as MapDiff and merge into Metadata#customs
+                combineClustersCustoms.set(
```
I'm probably missing something here, but just to double check: is it guaranteed that only one of these lambdas reaches the end? In other words, are we sure we won't encounter non-empty repositories for the default project _in both upserts and diffs_? I just want to make sure `combineClusterCustoms.set()` is only called once.

ywangd replied:
Yes, upserts and diffs are mutually exclusive, see 
https://github.com/elastic/elasticsearch/blob/dd58b0b6face632cd4476a496cb79e8c0dffdcb6/server/src/main/java/org/elasticsearch/cluster/DiffableUtils.java#L337-L348

nielsbauman reviewed (approved):
This is definitely not the most confident approval I've ever given. I agree with David and Pooya (and probably you do too) that this seralization logic is becoming extremely complex and unreadable. I'm happy some of this can be dropped in a few weeks already. I don't immediately have any better suggestions.

DaveCTurner reviewed (approved):
LGTM for a very loose definition of "good" which includes:

- this is at least well-covered by existing tests
- there are no _obvious_ mistakes here
- I cannot think of a better way of doing this given where we are today

ywangd committed:
Merge remote-tracking branch 'origin/main' into mp/repositories-metadata-migration

ywangd commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into mp/repositories-metadata-migration

ywangd commented:
Thanks all for the reviews! 🙏

ywangd committed:
Merge remote-tracking branch 'origin/main' into mp/repositories-metadata-migration
